toothpick-kotlin/src/main/java/toothpick/kotlin/ModuleDSL.kt

```diff
@@ -0,0 +1,76 @@
+package toothpick.kotlin
+
+import toothpick.config.Binding
+import toothpick.config.Module
+import javax.inject.Provider
+import kotlin.reflect.KClass
+
+fun module(block: ModuleBuilder.() -> Unit): Module = ModuleBuilder().apply(block).module
+
+class ModuleBuilder {
+    val module = Module()
+
+    inline infix fun <reified T : Any> KClass<T>.named(name: KClass<out Annotation>): Binding<T> {
+        return module.bind(T::class.java) named name
+    }
+
+    inline infix fun <reified T : Any> KClass<T>.named(name: String): Binding<T> {
+        return module.bind(T::class.java) named name
+    }
+
+    inline infix fun <reified T : Any> KClass<T>.providedAs(clazz: KClass<out T>): Binding<T>.BoundStateForClassBinding {
```

Comment:
all the fluent API has been renamed entirely. Can we keep the exact same method names for the bindings ?
I am not opposed to renaming them if needed, but it's a big change for TP. Are there any kotlin "way of doing things" that are at stake with these new names ?
Oh.. `to` is a kotlin keyword?

Reply:
I think the DSL should look more like:
```kotlin
bind T::class.java toInstance t0
bind T::class.java toClass T1::class.java
bind T::class.java withName "foo" toInstance t0
bind T::class.java toProviderInstance p0
bind T::class.java toProviderClass P::class.java

for `providesSingletonInScope`, `instancesInScope`, `singletonInScope`, I agree we have an issue that a infix method needs a param...
```

Reply:
Yeah, I went back and forth on this one a lot. Poked around in kodein and other kotlin DSLs to find a common naming scheme and didn't really see anything definitive. I settled on what I did as it read the most naturally to me as a human, but I'm not hugely attached to it. I like the examples you've provided.

Reply:
I also like the with style ... but really the main problem is that `to` is a reserved word.. I don't find it great I must say. 
Same goes for `as` , otherwise we could have had construct like
```
bind T::class.java to instance t0
bind T::class.java to instance t0 as singleton
```
This would be the more natural solution. 
Would it work if we would do
```
bind T::class.java `to` instance t0
bind T::class.java `to` instance t0 `as` singleton
```

It's ugly, but I am trying to see all our options.

Reply:
We could actually allow the to operator if module would support adding pairs..
```
module {
   bind Class<T> to instance(t0)
   bind Class<T> to Class<T1> 
   bind Class<T> to providerInstance(p0)
   bind Class<T> to Class<Provider<P1>>
}
```

Reply:
Interesting. I'll have to play with this to see if how it feels in real code. We're using a variant of the DSL internally so I can see how it goes with real world modules.